File: src/services/Catalogo/Catalogo.API/Models/ProdutoModel.cs
using Common.WebAPI.Validation;
using System.ComponentModel.DataAnnotations;

namespace Catalogo.API.Models
{
  public record ProdutoModel
  {
    [Required(ErrorMessage = "Nome do produto está vazio.")]
    public string? Nome { get; set; } = null!;

    [Required(ErrorMessage = "Descrição do produto está vazio.")]
    public string? Descricao { get; set; } = null!;

    [Required(ErrorMessage = "Preço do produto está vazio.")]
    [DecimalStringValidation(ErrorMessage = "Por favor, insira um número decimal válido.")]
    [Range(0.01, 9999999999999999.99, ErrorMessage = "O preço deve estar entre 0.01 e 9999999999999999.99.")]
    public decimal Preco { get; set; }

    [Required(ErrorMessage = "Unidade de medida do produto está vazio.")]
    public string? UnidadeMedida { get; set; }

    [Required(ErrorMessage = "Imagem do produto está vazio.")]
    public string? ImageUrl { get; set; }

    [Required(ErrorMessage = "Código de barras do produto está vazio.")]
    [CodigoDeBarrasProduto(ErrorMessage = "Código de barras inválido.")]
    public string? CodigoBarras { get; set; }

    [Required(ErrorMessage = "Estoque alvo do produto está vazio.")]
    [Range(0, 999, ErrorMessage = "O número deve estar entre 0 e 999.")]
    public int EstoqueAlvo { get; set; }

    public bool IsAtivo { get; set; }
  }

  public record ProdutoResponseModel
  {
    public string Id { get; set; }

    public ProdutoResponseModel(string id)
    {
      Id = id;
    }
  }
}

File: src/mobile/lib/app/utils/models/produtos/produto_model.dart
class ProdutoModel {
  ProdutoModel({
    required this.nome,
    required this.descricao,
    required this.imageUrl,
    required this.preco,
    required this.unidadeMedida,
    required this.codigoBarras,
    required this.estoqueAlvo,
    required this.isAtivo,
  });
  String? id;
  late final String nome;
  late final String descricao;
  late final String imageUrl;
  late final double preco;
  late final String unidadeMedida;
  late final String codigoBarras;
  late final int estoqueAlvo;
  late final bool isAtivo;

  Map<String, dynamic> toJson() {
    final data = <String, dynamic>{};
    data['nome'] = nome;
    data['descricao'] = descricao;
    data['imageUrl'] = imageUrl;
    data['preco'] = preco;
    data['unidadeMedida'] = unidadeMedida;
    data['codigoBarras'] = codigoBarras;
    data['estoqueAlvo'] = estoqueAlvo;
    data['isAtivo'] = isAtivo;
    return data;
  }

  ProdutoModel.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    nome = json['nome'];
    descricao = json['descricao'];
    imageUrl = json['imageUrl'];
    preco = json['preco'];
    unidadeMedida = json['unidadeMedida'];
    codigoBarras = json['codigoBarras'];
    estoqueAlvo = json['estoqueAlvo'];
    isAtivo = json['isAtivo'];
  }
}

class ProdutoResponseModel {
  ProdutoResponseModel({required this.id});
  late final String id;

  ProdutoResponseModel.fromJson(Map<String, dynamic> json) {
    id = json['id'];
  }
}
